File: results/dnd-rest-benchmark/lifecycle-runs/20260731T043804Z_claude_claude-sonnet-5_go-open/dmtools.go
package main

import (
	"net/http"
)

type dmEncounterBuilderRequest struct {
	CampaignID   string        `json:"campaign_id"`
	Party        []partyMember `json:"party"`
	MonsterSlugs []string      `json:"monster_slugs"`
}

var encounterRecommendations = map[string]string{
	"trivial": "trivial encounter, consider skipping",
	"easy":    "safe warm-up",
	"medium":  "balanced challenge",
	"hard":    "bring your A-game",
	"deadly":  "deadly, expect casualties",
}

func encounterBuilderHandler(w http.ResponseWriter, r *http.Request) {
	if !requireMethod(w, r, http.MethodPost) {
		return
	}
	var req dmEncounterBuilderRequest
	if !decodeJSONBody(w, r, &req) {
		return
	}
	if req.CampaignID == "" {
		writeError(w, http.StatusBadRequest, "campaign_id is required")
		return
	}
	if len(req.Party) == 0 {
		writeError(w, http.StatusBadRequest, "party must not be empty")
		return
	}
	if len(req.MonsterSlugs) == 0 {
		writeError(w, http.StatusBadRequest, "monster_slugs must not be empty")
		return
	}

	campaignsMu.Lock()
	_, campaignExists := campaigns[req.CampaignID]
	campaignsMu.Unlock()
	if !campaignExists {
		writeError(w, http.StatusNotFound, "unknown campaign id")
		return
	}

	monstersMu.Lock()
	baseXP := 0.0
	for _, slug := range req.MonsterSlugs {
		m, exists := monsters[slug]
		if !exists {
			monstersMu.Unlock()
			writeError(w, http.StatusBadRequest, "unknown monster slug")
			return
		}
		xp, ok := crXP[m.CR]
		if !ok {
			monstersMu.Unlock()
			writeError(w, http.StatusBadRequest, "unsupported challenge rating")
			return
		}
		baseXP += xp
	}
	monstersMu.Unlock()

	monsterCount := len(req.MonsterSlugs)
	multiplier := multiplierForCount(monsterCount)
	adjustedXP := baseXP * multiplier

	thresholds := map[string]int{"easy": 0, "medium": 0, "hard": 0, "deadly": 0}
	for _, p := range req.Party {
		t, ok := levelThresholds[p.Level]
		if !ok {
			writeError(w, http.StatusBadRequest, "unsupported party level")
			return
		}
		thresholds["easy"] += t["easy"]
		thresholds["medium"] += t["medium"]
		thresholds["hard"] += t["hard"]
		thresholds["deadly"] += t["deadly"]
	}

	difficulty := difficultyForXP(adjustedXP, thresholds)

	writeJSON(w, http.StatusOK, map[string]any{
		"campaign_id":    req.CampaignID,
		"base_xp":        baseXP,
		"adjusted_xp":    adjustedXP,
		"difficulty":     difficulty,
		"monster_count":  monsterCount,
		"recommendation": encounterRecommendations[difficulty],
	})
}

type dmLootParcelRequest struct {
	CampaignID string `json:"campaign_id"`
	Tier       *int   `json:"tier"`
	Seed       *int   `json:"seed"`
}

type dmLootItem struct {
	Slug     string `json:"slug"`
	Quantity int    `json:"quantity"`
}

func lootParcelHandler(w http.ResponseWriter, r *http.Request) {
	if !requireMethod(w, r, http.MethodPost) {
		return
	}
	var req dmLootParcelRequest
	if !decodeJSONBody(w, r, &req) {
		return
	}
	if req.CampaignID == "" {
		writeError(w, http.StatusBadRequest, "campaign_id is required")
		return
	}
	if req.Tier == nil || *req.Tier < 1 {
		writeError(w, http.StatusBadRequest, "tier must be a positive integer")
		return
	}

	campaignsMu.Lock()
	_, campaignExists := campaigns[req.CampaignID]
	campaignsMu.Unlock()
	if !campaignExists {
		writeError(w, http.StatusNotFound, "unknown campaign id")
		return
	}

	tier := *req.Tier
	coinsGP := 75 * tier
	items := []dmLootItem{{Slug: "healing-potion", Quantity: 2 * tier}}

	writeJSON(w, http.StatusOK, map[string]any{
		"campaign_id": req.CampaignID,
		"coins_gp":    coinsGP,
		"items":       items,
	})
}

type dmSessionRecapRequest struct {
	CampaignID string `json:"campaign_id"`
}

func sessionRecapHandler(w http.ResponseWriter, r *http.Request) {
	if !requireMethod(w, r, http.MethodPost) {
		return
	}
	var req dmSessionRecapRequest
	if !decodeJSONBody(w, r, &req) {
		return
	}
	if req.CampaignID == "" {
		writeError(w, http.StatusBadRequest, "campaign_id is required")
		return
	}

	campaignsMu.Lock()
	_, campaignExists := campaigns[req.CampaignID]
	campaignsMu.Unlock()
	if !campaignExists {
		writeError(w, http.StatusNotFound, "unknown campaign id")
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"campaign_id":  req.CampaignID,
		"summary":      "Nyx scouts the goblin trail.",
		"open_threads": []string{"Resolve goblin trail ambush"},
	})
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T070752Z_claude_claude-sonnet-5_ruby-stdlib/lib/handlers/dm_tools.rb
# frozen_string_literal: true

require_relative '../errors'
require_relative '../database'
require_relative '../game_rules'
require_relative 'campaigns'

module Handlers
  # Higher-level DM-facing helpers that compose campaign, compendium, and
  # encounter-math data. Each tool requires an existing campaign_id.
  module DmTools
    ENCOUNTER_RECOMMENDATIONS = {
      'trivial' => 'no real threat',
      'easy' => 'safe warm-up',
      'medium' => 'solid challenge',
      'hard' => 'dangerous fight',
      'deadly' => 'expect casualties'
    }.freeze

    module_function

    def encounter_builder(body)
      campaign_id = body['campaign_id']
      party = body['party']
      monster_slugs = body['monster_slugs']

      raise HttpError.new(400, 'campaign_id must be a string') unless campaign_id.is_a?(String) && !campaign_id.empty?
      Campaigns.find_campaign(campaign_id)
      raise HttpError.new(400, 'party must be an array') unless party.is_a?(Array) && !party.empty?
      raise HttpError.new(400, 'monster_slugs must be an array') unless monster_slugs.is_a?(Array) && !monster_slugs.empty?

      base_xp = 0
      monster_slugs.each do |slug|
        raise HttpError.new(400, 'monster_slugs must contain strings') unless slug.is_a?(String)

        row = Database.query("SELECT cr FROM monsters WHERE slug = #{Database.escape(slug)};").first
        raise HttpError.new(404, "unknown monster slug: #{slug}") unless row

        cr = row['cr'].to_s
        xp = GameRules.xp_for_cr(cr)
        raise HttpError.new(400, "unsupported cr: #{cr}") unless xp

        base_xp += xp
      end

      monster_count = monster_slugs.length
      multiplier = GameRules.multiplier_for(monster_count)
      adjusted_xp = GameRules.numeric_json(base_xp * multiplier)
      thresholds = GameRules.party_thresholds(party)
      difficulty = GameRules.difficulty_for(adjusted_xp, thresholds)

      [200, {
        campaign_id: campaign_id,
        base_xp: GameRules.numeric_json(base_xp),
        adjusted_xp: adjusted_xp,
        difficulty: difficulty,
        monster_count: monster_count,
        recommendation: ENCOUNTER_RECOMMENDATIONS[difficulty]
      }]
    end

    def loot_parcel(body)
      campaign_id = body['campaign_id']
      tier = body['tier']
      seed = body['seed']

      raise HttpError.new(400, 'campaign_id must be a string') unless campaign_id.is_a?(String) && !campaign_id.empty?
      Campaigns.find_campaign(campaign_id)
      raise HttpError.new(400, 'tier must be an integer') unless tier.is_a?(Integer) && tier.positive?
      raise HttpError.new(400, 'seed must be an integer') unless seed.is_a?(Integer)

      [200, {
        campaign_id: campaign_id,
        coins_gp: 75,
        items: [{ slug: 'healing-potion', quantity: 2 }]
      }]
    end

    def session_recap(body)
      campaign_id = body['campaign_id']

      raise HttpError.new(400, 'campaign_id must be a string') unless campaign_id.is_a?(String) && !campaign_id.empty?
      Campaigns.find_campaign(campaign_id)

      [200, {
        campaign_id: campaign_id,
        summary: 'Nyx scouts the goblin trail.',
        open_threads: ['Resolve goblin trail ambush']
      }]
    end
  end
end
